### ISSUE TITLE
Bug in parallel_reduce with HIP

### ISSUE THREAD

**Issue by stanmoore1** [kokkos/kokkos#3442]
I'm seeing a bug in `parallel_reduce` with the HIP backend, where the final sum value is wrong. It seems to be related to using a large `value_type` for the reduction. This large `value_type` is from a real use case in LAMMPS.

Here is a small reproducer: 
[bug.cpp.txt](https://github.com/kokkos/kokkos/files/5309184/bug.cpp.txt)

On an MI60 AMD GPU it fails:
```
Sum of squares of integers from 0 to 299, computed in parallel, is 20480
Sum of squares of integers from 0 to 299, computed sequentially, is 8955050
```

On a CPU it works correctly:
```
Sum of squares of integers from 0 to 299, computed in parallel, is 8955050
Sum of squares of integers from 0 to 299, computed sequentially, is 8955050
```

**Comment by stanmoore1** [comment]
@Rombur

**Comment by Rombur** [comment]
It looks like there is a register spill which makes sense since you have a large `value_type`. This triggers a bug in `hipcc`. AMD is aware of the bug but there is no timeline on when this will be fixed. BTW, this is the same problem than in TestSNAP, it just a different outcome (wrong result instead of crashing).

**Comment by stanmoore1** [comment]
It looks like the largest `value_type` that doesn't fail is 8 doubles. I need at least 11 doubles to be able to run LAMMPS.

**Comment by arghdos** [comment]
I'll take a look today.

**Comment by stanmoore1** [comment]
A (temporary) workaround would be to take 3 doubles out of the `value_type` and use atomic-adds to do a global sum.

**Comment by Rombur** [comment]
@stanmoore1 `double` are fines but there is a bug with `complex` atomic in Kokkos. I am working on it but right now I would not advise you to use `complex` atomic.

**Comment by arghdos** [comment]
Through the 'magic' of launch bounds, I have managed to make the sum zero:

>Sum of squares of integers from 0 to 299, computed in parallel, is 0
Sum of squares of integers from 0 to 299, computed sequentially, is 8955050

:stuck_out_tongue: 

It does look related to the issues we've been having in LAMMPS.  The entire ev struct is being spilled to HBM here.  Will be a good reproducer for our compiler teams :)

**Comment by arghdos** [comment]
With @Rombur's recent PR #3467, I (seemingly) have a workaround for this in ROCm 3.8 -- if you template the structs based on the size, e.g.,:

```
template<int vsize, int esize>
struct t_EV_FLOAT_REAX {
  double evdwl;
  double ecoul;
  double v[vsize];
  double ereax[esize];
  KOKKOS_INLINE_FUNCTION
  t_EV_FLOAT_REAX() {
    evdwl = 0;
    ecoul = 0;
    for (int i = 0; i < vsize; ++i) v[i] = 0;
    for (int i = 0; i < esize; ++i) ereax[i] = 0;
  }

  KOKKOS_INLINE_FUNCTION
  void operator+=(const t_EV_FLOAT_REAX<vsize, esize> &rhs) {
    evdwl += rhs.evdwl;
    ecoul += rhs.ecoul;
    for (int i = 0; i < vsize; ++i) v[i] += rhs.v[i];
    for (int i = 0; i < esize; ++i) ereax[i] += rhs.ereax[i];
  }

  KOKKOS_INLINE_FUNCTION
  void operator+=(const volatile t_EV_FLOAT_REAX<vsize, esize> &rhs) volatile {
    evdwl += rhs.evdwl;
    ecoul += rhs.ecoul;
    for (int i = 0; i < vsize; ++i) v[i] += rhs.v[i];
    for (int i = 0; i < esize; ++i) ereax[i] += rhs.ereax[i];
  }
};
typedef struct t_EV_FLOAT_REAX<6, 10> tEV_FLOAT_REAX;
```

you get correct answers:

```
Sum of squares of integers in native REAX from 0 to 299, computed in parallel, is 1113775
Sum of squares of integers in templated REAX from 0 to 299, computed in parallel, is 8955050
Sum of squares of integers in templated small REAX from 0 to 299, computed in parallel, is 8955050
Sum of squares of integers from 0 to 299, computed sequentially, is 8955050
```

[reducebug.cpp.txt](https://github.com/kokkos/kokkos/files/5365841/reducebug.cpp.txt)

**Comment by stanmoore1** [comment]
@Rombur that is a good workaround, when I get a chance I put that in a LAMMPS branch and try it out for ReaxFF.

**Comment by stanmoore1** [comment]
Whoops I meant to tag @arghdos

**Comment by arghdos** [comment]
Figured out the cause -- turns out that the EV_FLOAT_REAX type declares an ereax array of size 10 here:

https://github.com/lammps/lammps/blob/master/src/KOKKOS/kokkos_type.h#L408  

however, ereax[9] is never referenced or used, so the compiler seems to optimize it out.  Changing this to either use:

```
double ereax[9];  
```

or adding:

```
ereax[9] = 0;
...
ereax[9] += rhs.ereax[9];
...
ereax[9] += rhs.ereax[9];
```

resolves the issue on ROCm 3.9.  From the Kokkos side, I think this is resolved w/ @Rombur's race fix.

@stanmoore1 -- is ereax[9] used anywhere?

**Comment by stanmoore1** [comment]
@arghdos somehow I missed your last comment. I fixed the unused variable in https://github.com/lammps/lammps/pull/2485.

**Comment by arghdos** [comment]
This should be fixed in 4.2

**Comment by stanmoore1** [comment]
Great!

**Comment by stanmoore1** [comment]
I confirmed this is fixed in ROCm 4.2. Closing #3442.
